### PR TITLE
refactor: yarn constraint determination

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Refactors and adds a test for yarn constraint determination.

## Context

Preparing for #27656

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Apparently we fall back to Yarn 2 right now: https://github.com/renovatebot/renovate/blob/66db519d3707e606e5849e521b58dec0197093e4/lib/modules/manager/npm/extract/yarn.ts#L102-L116

File: lib/modules/manager/npm/post-update/yarn.spec.ts [hunk 1]
```diff
@@ -409,6 +409,47 @@ describe('modules/manager/npm/post-update/yarn', () => {
     expect(res.lockFile).toBe('package-lock-contents');
   });
 
+  it('falls back to Yarn v1 for undefined constraint', async () => {
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/yarn.spec.ts:412]: Apparently we default to Yarn v2 due to this: https://github.com/renovatebot/renovate/blob/66db519d3707e606e5849e521b58dec0197093e4/lib/modules/manager/npm/extract/yarn.ts#L102-L116

File: lib/modules/manager/npm/post-update/yarn.ts [hunk 2]
```diff
@@ -90,6 +90,41 @@ export function isYarnUpdate(upgrade: Upgrade): boolean {
...
+  const lockFileName = upath.join(lockFileDir, 'yarn.lock');
+  const lockConstraint = getYarnVersionFromLock(
+    await getYarnLock(lockFileName),
+  );
+  if (lockConstraint) {
+    logger.debug('Using yarn constraint from yarn.lock');
+    return lockConstraint;
+  }
+  logger.debug('No yarn constraint found');
+  return undefined;
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/yarn.ts:125]: These lines are unreachable due to https://github.com/renovatebot/renovate/blob/66db519d3707e606e5849e521b58dec0197093e4/lib/modules/manager/npm/extract/yarn.ts#L102-L116

File: lib/modules/manager/npm/post-update/yarn.ts [hunk 3]
```diff
@@ -90,6 +90,41 @@ export function isYarnUpdate(upgrade: Upgrade): boolean {
...
+  lockFileDir: string,
+  config: Partial<PostUpdateConfig<NpmManagerData>>,
+  upgrades: Upgrade[],
+): Promise<string | undefined> {
+  const yarnUpdate = upgrades.find(isYarnUpdate);
+  if (yarnUpdate?.newValue) {
+    logger.debug('Using yarn update newValue as constraint');
+    return yarnUpdate.newValue;
+  }
+  if (config.constraints?.yarn) {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/yarn.ts:103]: didn't we agree to use this first?

File: lib/modules/manager/npm/post-update/yarn.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/manager/npm/post-update/yarn.ts:125]: so simply return lock constraint?

**Comment by viceice** [comment]
we should latest yarn if no constraints can be found.

**Pull request closed without merging**